src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/DeclareCommands.java

```diff
@@ -0,0 +1,475 @@
+
+	// PE argument types that we use. These do not match values used by Nukkit, since they did
+	// not work for us. The reason for this is unclear. Many different values produced seemingly
+	// the same data type in the PE client chat GUI. Documented here for posterity:
+	// 8-13 target (with wildcard)
+	// 14-17 file path
+	// 18-26 "unknown"
+	// 27-28 string
+	// 29-31 position ("x y z")
+	// 32-33 message
+	// 34-36 text
+	// 37-43 json
+	// 44-... ? (tested up to 99) command
+	public static final int ARG_TYPE_INT = 1;
+	public static final int ARG_TYPE_FLOAT = 2;
+	public static final int ARG_TYPE_TARGET = 6;
+	public static final int ARG_TYPE_STRING = 27;
+	public static final int ARG_TYPE_POSITION = 29;
+	public static final int ARG_TYPE_MESSAGE = 32;
+	// Not sure if we will need these..?
+	public static final int ARG_TYPE_RAWTEXT = 34;
+	public static final int ARG_TYPE_JSON = 37;
+	public static final int ARG_TYPE_COMMAND = 44;
+
+	public static final int ARG_FLAG_VALID = 0x100000;
+	public static final int ARG_FLAG_LITERAL = 0x200000;
+
+	private CommandNode[] allNodes;
+	private CommandNode[] topLevelNodes; // A subset of allNodes containing the starting "command" nodes
+	private int rootNodeIndex;
+
+	private static class CommandNode {
+		private final String name;
+		private final String argType;
+		private final String suggestion;
+		private final int[] children;
+		private final int redirect;
+		private final boolean isPathEnd;
+
+		public CommandNode(String name, String argType, String suggestion, int[] children, int redirect, boolean isPathEnd) {
+			this.name = name;
+			this.argType = argType;
+			this.suggestion = suggestion;
+			this.children = children;
+			this.redirect = redirect;
+			this.isPathEnd = isPathEnd;
+		}
+
+		@Override
+		public String toString() {
+			return Utils.toStringAllFields(this);
+		}
+
+		public boolean isLeaf() {
+			return this.getChildren().length == 0;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public String getArgType() {
+			return argType;
+		}
+
+		public int[] getChildren() {
+			return children;
+		}
+	}
+
+	private static class PEArgumentNode {
+		private final String name;
+		private final String argType;
+		private final int nameIndex;
+
+		public PEArgumentNode(CommandNode node, PECommandsStructure peStruct) {
+			this.name = node.getName();
+			this.argType = node.getArgType();
+
+			// Cache literal index, if this is a literal argument
+			if (this.argType == null) {
+				this.nameIndex = peStruct.registerLiteral(this.name);
+			} else {
+				this.nameIndex = -1;
+			}
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public String getArgType() {
+			return argType;
+		}
+
+		public int getNameIndex() {
+			return nameIndex;
+		}
+	}
+
+	private static class PECommandsStructure {
+		private Map<CommandNode, List<List<PEArgumentNode>>> overloadsRegistry = new HashMap<>();
+		private Map<String, Integer> literalRegistry = new HashMap<>();
```

Comment:
This can be just a Set based on LinkedHashMap. That way you don't have to maintain an index. Will also allow just passing Collection to writeLiteral methods.

Reply:
I'm not sure I follow you 100%. My main constraint here is that I need to assign each and every unique literal a value; these values must be consecutive, and I must at a later point be able to do two-way conversions (i.e. string->num and num->string). I also prefer to do these conversations as efficient as possible.

I can't really see how a Set can do that.

However, for the overloads, they could be Set<List<PEArgumentNode>> since the order between overloads do not (or should not) matter, only between arguments in each overload. However, I prefer to keep them in order; I have not tested but presumably the PE client is brain dead and presents them in the order I send them.

Reply:
Based on the code i seem, you don't need string->num and num->string convert.
With linkedhashmap the iteration order index will be your index (which can be used in writeLiteral*) and you can just add the new literal to the map if it is not already present, so the order will stay.

Reply:
Yes, it might be possible to use a LinkedHashMap instead. I'd forgotten about that data type. However, I'm not sure how much it improves the code. We'll still need to use the Map part of it to map String to Integer, mapping in an increasing value as is done now, and then I can iterate over the List part in insertion order -- this will happen to correspond to the mapped Integers, but this will not be due to the design of the datatype, but due to careful usage.

Ah heck. I should probably subclass LinkedHashMap and make this functionality explicit.

Reply:
Yeah, the usage of LinkedHashMap in such manner will require a comment about that. That actually decrease code readability, yet improves performance.